superset/utils/core.py

```diff
@@ -1401,19 +1407,48 @@ def get_column_names_from_metrics(metrics: List[Metric]) -> List[str]:
     return columns
 
 
-def serialize_pandas_dtypes(dtypes: List[np.dtype]) -> List[GenericDataType]:
-    """Serialize pandas/numpy dtypes to JavaScript types"""
-    mapping = {
-        "object": GenericDataType.STRING,
-        "category": GenericDataType.STRING,
-        "datetime64[ns]": GenericDataType.TEMPORAL,
-        "int64": GenericDataType.NUMERIC,
-        "in32": GenericDataType.NUMERIC,
-        "float64": GenericDataType.NUMERIC,
-        "float32": GenericDataType.NUMERIC,
-        "bool": GenericDataType.BOOLEAN,
+def extract_dataframe_dtypes(df: pd.DataFrame) -> List[GenericDataType]:
+    """Serialize pandas/numpy dtypes to generic types"""
+
+    def _check_dtype(series: pd.Series) -> Optional[GenericDataType]:
+        if is_numeric_dtype(series):
+            return GenericDataType.NUMERIC
+        elif is_bool_dtype(series):
+            return GenericDataType.BOOLEAN
+        elif is_datetime64_any_dtype(series):
+            return GenericDataType.TEMPORAL
+        return None
+
+    type_map: Dict[Type[Any], GenericDataType] = {
+        date: GenericDataType.TEMPORAL,
+        datetime: GenericDataType.TEMPORAL,
+        pd.Timestamp: GenericDataType.TEMPORAL,
+        pd.DatetimeTZDtype: GenericDataType.TEMPORAL,
+        str: GenericDataType.STRING,
+        int: GenericDataType.NUMERIC,
+        float: GenericDataType.NUMERIC,
+        bool: GenericDataType.BOOLEAN,
     }
-    return [mapping.get(str(x), GenericDataType.STRING) for x in dtypes]
+
+    generic_dtypes: List[GenericDataType] = []
+    for idx, column in enumerate(df.columns):
+        generic_dtype = GenericDataType.STRING
+        series = df[column]
+        if is_object_dtype(df[column]):
```

Comment:
Have you tried [`pandas.api.types.infer_dtype`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.api.types.infer_dtype.html)?